twins/api/http/endpoint_states_test.go

```diff
@@ -0,0 +1,259 @@
+			auth:   token,
+			status: http.StatusBadRequest,
+			url:    fmt.Sprintf("%s?offset=invalid&limit=%d", baseURL, 15),
+			res:    nil,
+		},
+		{
+			desc:   "get a list of states with invalid limit",
+			auth:   token,
+			status: http.StatusBadRequest,
+			url:    fmt.Sprintf("%s?offset=%d&limit=invalid", baseURL, 0),
+			res:    nil,
+		},
+		{
+			desc:   "get a list of states without offset",
+			auth:   token,
+			status: http.StatusOK,
+			url:    fmt.Sprintf("%s?limit=%d", baseURL, 15),
+			res:    data[0:15],
+		},
+		{
+			desc:   "get a list of states without limit",
+			auth:   token,
+			status: http.StatusOK,
+			url:    fmt.Sprintf("%s?offset=%d", baseURL, 14),
+			res:    data[14:24],
+		},
+		{
+			desc:   "get a list of states with invalid number of parameters",
+			auth:   token,
+			status: http.StatusBadRequest,
+			url:    fmt.Sprintf("%s%s", baseURL, "?offset=4&limit=4&limit=5&offset=5"),
+			res:    nil,
+		},
+		{
+			desc:   "get a list of states with redundant query parameters",
+			auth:   token,
+			status: http.StatusOK,
+			url:    fmt.Sprintf("%s?offset=%d&limit=%d&value=something", baseURL, 0, 5),
+			res:    data[0:5],
+		},
+	}
+
+	for _, tc := range cases {
+		req := testRequest{
+			client: ts.Client(),
+			method: http.MethodGet,
+			url:    tc.url,
+			token:  tc.auth,
+		}
+		res, err := req.make()
+		assert.Nil(t, err, fmt.Sprintf("%s: unexpected error %s", tc.desc, err))
+
+		var resData statesPageRes
+		if tc.res != nil {
+			err = json.NewDecoder(res.Body).Decode(&resData)
+			assert.Nil(t, err, fmt.Sprintf("%s: unexpected error %s", tc.desc, err))
+		}
+
+		assert.Equal(t, tc.status, res.StatusCode, fmt.Sprintf("%s: expected status code %d got %d", tc.desc, tc.status, res.StatusCode))
+		assert.ElementsMatch(t, tc.res, resData.States, fmt.Sprintf("%s: expected body %v got %v", tc.desc, tc.res, resData.States))
+	}
+}
+
+func createAttribute(name, subtopic string) twins.Attribute {
+	id, _ := uuid.New().ID()
+	return twins.Attribute{
+		Name:         name,
+		Channel:      id,
+		Subtopic:     subtopic,
+		PersistState: true,
+	}
+}
+
+func createSenML(n int, bn string) []senml.Record {
+	var recs []senml.Record
+	for i := 0; i < n; i++ {
+		rec := senml.Record{
+			BaseName: bn,
+			BaseTime: float64(time.Now().Unix()),
+			Time:     float64(i),
+			Value:    nil,
+		}
+		recs = append(recs, rec)
+	}
+	return recs
+}
+
+func createMessage(attr twins.Attribute, recs []senml.Record) (*messaging.Message, error) {
+	mRecs, err := json.Marshal(recs)
+	if err != nil {
+		return nil, err
+	}
+	return &messaging.Message{
+		Channel:   attr.Channel,
+		Subtopic:  attr.Subtopic,
+		Payload:   mRecs,
+		Publisher: publisher,
+	}, nil
+}
+
+func createStateResponse(id int, tw twins.Twin, rec senml.Record) stateRes {
+	return stateRes{
+		TwinID:     tw.ID,
+		ID:         int64(id),
+		Definition: tw.Definitions[len(tw.Definitions)-1].ID,
+		Payload:    map[string]interface{}{rec.BaseName: nil},
```

Comment:
I don't think that you need to set this to nil, zero value of pointer is `nil`

Reply:
I have to initialize this map to `{rec.BaseName: nil}`, therefore, I need to put explicit values.

Reply:
ok, its map init, my bad